### PR TITLE
test: unit test suite (210 tests) + SonarQube CI integration

### LINKED ISSUE #267: Tests
- [ ] Unit Tests (Front & Back)
- [ ] Integration Tests (Front & Back)
- [ ] E2E Tests (Front & Back)
- [ ] Sonarqube

### PR DESCRIPTION
**Pull request by Xayrulloh**
## Summary

- Added **210 unit tests** across 5 tiers: pure helpers, Zod schemas, NestJS guards, delete-preview logic, and web Effector factories
- Set up Jest for `apps/api` and Vitest for `libs/shared` with full test infrastructure (`jest.config.ts`, `vitest.config.ts`, `tsconfig.spec.json`)
- Applied AAA (Arrange-Act-Assert) blank-line formatting across all 21 spec files; added `/// <reference types="jest" />` to all API specs for reliable IDE type resolution
- Added `biome.json` overrides for `*.spec.ts` disabling `noExplicitAny` and `noNonNullAssertion` — safe in test mock contexts
- Wired SonarQube analysis into CI: `lcov` coverage reporters in all 3 test runners, `sonar-project.properties` per package, and `nx affected -t scan` step in the GitHub Actions workflow

## Changes

| Area | Files |
|---|---|
| API tests | `apps/api/src/common/guards/*.spec.ts` (4 files), `apps/api/src/modules/family-tree-member/services/*.spec.ts`, `apps/api/src/helpers/random-string.helper.spec.ts` |
| Web tests | `apps/web/src/shared/lib/*.spec.ts` (4 files) |
| Shared tests | `libs/shared/src/lib/schema/*.spec.ts` (10 files), `libs/shared/src/lib/helpers/random-avatar.spec.ts` |
| Test infra | `apps/api/jest.config.ts`, `apps/api/tsconfig.spec.json`, `libs/shared/vitest.config.ts`, `libs/shared/tsconfig.spec.json`, `libs/shared/tsconfig.json` |
| SonarQube | `apps/api/sonar-project.properties`, `apps/web/sonar-project.properties`, `libs/shared/sonar-project.properties`, `apps/web/vite.config.ts`, `biome.json` |
| CI | `.github/workflows/ci.yml` |

## Test plan

- [ ] `npx nx test api` → 46 tests pass (6 suites)
- [ ] `npx nx test shared` → 103 tests pass (11 suites)
- [ ] `npx nx test web` (from `apps/web/`) → 52 tests pass (4 suites)
- [ ] `npx nx test api --coverage` → `coverage/apps/api/lcov.info` generated
- [ ] `npx nx test shared --coverage` → `coverage/libs/shared/lcov.info` generated
- [ ] `npx vitest run --coverage` (from `apps/web/`) → `coverage/apps/web/lcov.info` generated
- [ ] CI passes on PR open (lint + build + tests with coverage + SonarQube scan)
- [ ] Add `SONAR_TOKEN` secret to the repo before merging if not already set

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Tests**
  * Added extensive unit test coverage for backend authorization guards, shared schemas/helpers, and web UI utilities.
  * Added service-level tests for deletion preview logic and member relationship edge cases.
* **Chores**
  * Enhanced CI to run unit tests with coverage and perform automated code analysis.
  * Updated Jest/Vitest and Sonar configurations to improve test discovery and coverage reporting.
* **Refactor**
  * Updated “time ago” formatting to use consistent pluralized “n unit(s) ago” output.

### CODE REVIEW

File: apps/api/src/modules/family-tree-member/services/family-tree-member.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+  let service: FamilyTreeMemberService;
+
+  beforeEach(() => {
+    jest.clearAllMocks();
+    service = new FamilyTreeMemberService(
+      mockDb as any,
+      {} as any,
+      { getOrThrow: () => 'https://r2.example.com' } as any,
+    );
+  });
```
[the remark this answers is not in the document]
Comment (Xayrulloh) [reply apps/api/src/modules/family-tree-member/services/family-tree-member.service.spec.ts:99]: testing computeDeletePreview via (service as any) was intentional; the method is complex private logic with no public entry point. Comment: "intentional — private method has no public API surface; testing directly is a deliberate tradeoff"

File: apps/web/src/shared/lib/create-form.spec.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      expect($formValues.getState()).toEqual({});
+    });
+  });
+
+  describe('effects — throw when form is not bound', () => {
+    it('clearErrorsFx throws when $formInstance is null', async () => {
+      const { clearErrorsFx } = createForm<TestShape>();
+
+      await expect(clearErrorsFx()).rejects.toThrow('Form instance is not initialized');
+    });
```
[the remark this answers is not in the document]
Comment (Xayrulloh) [reply apps/web/src/shared/lib/create-form.spec.ts:41]: exact messages are the contract; assert-by-type would lose specificity

File: apps/web/src/shared/lib/create-form.spec.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        'Form instance is not initialized',
+      );
+    });
+  });
+
+  describe('useBindFormWithModel + effects', () => {
+    it('populates $formInstance when the hook mounts', () => {
+      const { $formInstance, useBindFormWithModel } = createForm<TestShape>();
+      const mockForm = makeMockForm();
+      renderHook(() => useBindFormWithModel({ form: mockForm as unknown as MockFormInstance }));
```
[the remark this answers is not in the document]
Comment (Xayrulloh) [reply apps/web/src/shared/lib/create-form.spec.ts]: standard typed mock pattern, no runtime risk

File: apps/web/src/shared/lib/family-chart-transformer.spec.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+  });
+
+  describe('SPOUSE connections', () => {
+    it('creates bidirectional spouse links', () => {
+      const members = [makeMember('a'), makeMember('b', 'FEMALE')];
+      const connections = [makeConnection('a', 'b', FamilyTreeMemberConnectionEnum.SPOUSE)];
+
+      const result = toF3Data(members, connections);
+      const a = result.find((d) => d.id === 'a')!;
+      const b = result.find((d) => d.id === 'b')!;
```
[the remark this answers is not in the document]
Comment (Xayrulloh) [reply apps/web/src/shared/lib/family-chart-transformer.spec.ts:101]: non-null assertion in specs is permitted by project Biome override; find result is guaranteed by test setup

**Comment by Xayrulloh** [comment]
@coderabbitai pls review again and if all LGTY feel free to approve it)

**Comment by Xayrulloh** [comment]
https://github.com/coderabbitai pls review again and if all LGTY feel free to approve it)

**Comment by Xayrulloh** [comment]
@coderabbitai pls review again and if all LGTY feel free to approve it)
